### PR TITLE
perf: add 3-layer early date filtering to skip irrelevant files and entries

### PR DESCRIPTION
**Pull request by pbuchman**
## Summary

- Add file-level mtime pre-filter to skip JSONL files that can't contain entries for the requested date range
- Add line-level substring pre-check (`USAGE_LINE_MARKER`) to skip `JSON.parse` on non-usage lines
- Add entry-level early date skip to avoid cost calculation for out-of-range entries
- Apply optimizations to `loadDailyUsageData`, `loadSessionData`, and `loadSessionBlockData`

## Problem

When `--since` / `--until` are provided, ccusage currently:
1. Globs **all** JSONL files
2. Opens **every** file to extract its earliest timestamp (for sorting)
3. Reads and parses **every** line of **every** file
4. Calculates cost for **every** entry
5. Groups and aggregates **all** entries
6. **Then** filters by date range — discarding most of the work

This means a single-day query takes the same time as a full scan.

## Benchmarks

**Machine:** Apple M2 Pro, 32 GB RAM, macOS 15.6
**Dataset:** 4,637 JSONL files, 572 MB total (largest file: 469 MB)

| Query | Before | After | Improvement |
|-------|--------|-------|-------------|
| Today only (`--since 20260227 --until 20260227`) | 36.7s | 0.76s | **98%** |
| Last 7 days (`--since 20260220`) | 36.7s | 3.0s | **92%** |
| Full scan (no filters) | 36.7s | 31.6s | **14%** |

The mtime filter provides the biggest wins for short date ranges (the most common use case for integrations like OpenUsage). The substring pre-check also improves full-scan performance by ~14% by skipping `JSON.parse` on non-usage lines.

### 1. File-level mtime pre-filter (`filterFilesByMtime`)
Before reading any file content, checks each file's `mtime` against the `--since` date (with 1-day buffer for timezone safety). Files whose `mtime` is older than `since - 1 day` are skipped entirely.

### 2. Line-level substring pre-check (`USAGE_LINE_MARKER`)
Before calling `JSON.parse` on each line, checks whether the line contains the `"costUSD"` substring (the `USAGE_LINE_MARKER`). In typical JSONL files, ~91.5% of lines are non-usage entries (conversation turns, system messages, etc.) that don't contain cost data. This skips expensive JSON parsing for those lines entirely.

### 3. Entry-level early date skip
Inside the per-line processing loop, after `formatDate` but **before** the expensive `calculateCostForEntry`, compares the entry's date against `since`/`until` and skips out-of-range entries. Uses the same comparison logic as the existing `filterByDateRange` in `_date-utils.ts`.

## Backwards compatibility

This is a **backwards-compatible** change:
- mtime filtering uses a 1-day buffer to account for timezone edge cases
- Substring pre-check only skips lines that can't possibly be usage entries
- Entry-level skip uses identical comparison logic to the existing `filterByDateRange`
- No behavioral change when `--since`/`--until` are not provided
- Files that fail `stat()` are kept (not filtered), ensuring no data loss on permission issues
- All existing tests continue to pass

## Context

Discovered while debugging slow load times in [OpenUsage](https://github.com/robinebers/openusage), a menu bar app that uses ccusage to display Claude Code token usage. With 15-second timeouts per runner, ccusage always times out on machines with heavy usage, causing 60+ second delays. See: robinebers/openusage#249

## Testing

- Added unit tests for `parseSinceDate` (YYYYMMDD parsing, edge cases)
- Added unit tests for `filterFilesByMtime` (mtime edge cases, stat errors, buffer behavior)
- Added tests for `USAGE_LINE_MARKER` substring pre-check (true/false positives, edge cases)
- Added tests for entry-level early skip (correctness parity with existing post-filter)
- Added integration tests with multi-date fixtures and mtime simulation via `utimes()`
- All 269 tests pass: `pnpm run format && pnpm typecheck && pnpm run test`

## Summary by CodeRabbit

* **New Features**
  * Date-range (since/until) support added to data loading options.

* **Improvements**
  * Pre-filters files by modification time and applies date-range skipping earlier to reduce work and speed processing.
  * Skips irrelevant lines before parsing and ensures deduplication and cost calculations respect date filters and timezones.

* **Tests**
  * Expanded coverage for date parsing, mtime filtering, line-level skipping, and end-to-end date/time behavior.

### CODE REVIEW

File: apps/ccusage/src/data-loader.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@ import type { LoadedUsageEntry, SessionBlock } from './_session-blocks.ts';
 import type { ActivityDate, Bucket, CostMode, ModelName, SortOrder, Version } from './_types.ts';
 import { Buffer } from 'node:buffer';
 import { createReadStream, createWriteStream } from 'node:fs';
-import { readFile } from 'node:fs/promises';
+import { readFile, stat } from 'node:fs/promises';
```
[the remark this answers is not in the document]
Comment (pbuchman) [reply apps/ccusage/src/data-loader.ts]: Fixed in 33eac4f — removed both `await import('fs-fixture')` calls and the `async` from the describe callbacks. `createFixture` was already imported at the top level (line 23).

File: apps/ccusage/src/data-loader.ts [hunk 2]
```diff
@@ -716,6 +716,41 @@ export async function globUsageFiles(claudePaths: string[]): Promise<GlobResult[
 	return (await Promise.all(filePromises)).flat();
 }
 
+/**
+ * Parse a YYYYMMDD string into a Date object (local time, midnight)
+ */
+export function parseSinceDate(since: string): Date {
+	const y = Number(since.slice(0, 4));
```
[the remark this answers is not in the document]
Comment (pbuchman) [reply apps/ccusage/src/data-loader.ts]: Already addressed in v2 — both `parseSinceDate` and `filterFilesByMtime` are not exported (plain `function`/`async function`).

File: apps/ccusage/src/data-loader.ts [hunk 3]
```diff
@@ -716,6 +716,41 @@ export async function globUsageFiles(claudePaths: string[]): Promise<GlobResult[
...
+ * Uses a 1-day buffer before the since date to account for timezone edge cases.
+ */
+export async function filterFilesByMtime(files: string[], since?: string): Promise<string[]> {
+	if (since == null) {
+		return files;
+	}
+
+	const sinceDate = parseSinceDate(since);
+	sinceDate.setDate(sinceDate.getDate() - 1);
+	const sinceMs = sinceDate.getTime();
```
[the remark this answers is not in the document]
Comment (pbuchman) [reply apps/ccusage/src/data-loader.ts]: Already addressed in v2 — the guard now checks `since == null || since.trim() === '' || \!/^\\d{8}$/.test(since)`, covering null, empty string, and invalid format.

File: apps/ccusage/src/data-loader.ts [hunk 4]
```diff
@@ -976,6 +1033,12 @@ export async function loadSessionData(options?: LoadOptions): Promise<SessionUsa
 				markAsProcessed(uniqueHash, processedHashes);
 
 				const sessionKey = `${projectPath}/${sessionId}`;
+
+				const entryDate = formatDate(data.timestamp, options?.timezone, DEFAULT_LOCALE);
+				if (isOutsideDateRange(entryDate, options?.since, options?.until)) {
+					return;
+				}
```
[the remark this answers is not in the document]
Comment (pbuchman) [reply apps/ccusage/src/data-loader.ts]: Fixed in 33eac4f — removed entry-level date skip from `loadSessionData`. Sessions are now filtered post-aggregation by `lastActivity`, preserving correct totals for sessions spanning date boundaries.

File: apps/ccusage/src/data-loader.ts [hunk 5]
```diff
@@ -1402,6 +1468,11 @@ export async function loadSessionBlockData(options?: LoadOptions): Promise<Sessi
 				// Mark this combination as processed
 				markAsProcessed(uniqueHash, processedHashes);
 
+				const entryDate = formatDate(data.timestamp, options?.timezone, DEFAULT_LOCALE);
+				if (isOutsideDateRange(entryDate, options?.since, options?.until)) {
+					return;
+				}
```
[the remark this answers is not in the document]
Comment (pbuchman) [reply apps/ccusage/src/data-loader.ts]: Fixed in 33eac4f — removed entry-level date skip from `loadSessionBlockData`. Blocks are now constructed from the full entry stream and filtered post-construction by `block.startTime`, preserving correct block boundaries.

File: apps/ccusage/src/data-loader.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pbuchman) [reply apps/ccusage/src/data-loader.ts]: Already present — line 16 reads `import { readFile, stat, utimes } from 'node:fs/promises';`. The v2 rewrite included `utimes` from the start.

File: apps/ccusage/src/data-loader.ts [hunk 6]
```diff
@@ -716,6 +716,51 @@ export async function globUsageFiles(claudePaths: string[]): Promise<GlobResult[
...
+}
+
+async function filterFilesByMtime(files: string[], since?: string): Promise<string[]> {
+	if (since == null || since.trim() === '' || !/^\d{8}$/.test(since)) {
+		return files;
+	}
+
+	const sinceDate = parseSinceDate(since);
+	sinceDate.setDate(sinceDate.getDate() - 1);
+	const sinceMs = sinceDate.getTime();
```
[the remark this answers is not in the document]
Comment (pbuchman) [reply apps/ccusage/src/data-loader.ts:743]: Fixed in 4cf1a24 — added calendar date validation. Impossible dates like `20230231` or `20231301` now bypass the mtime pre-filter instead of applying a rolled-over cutoff. Added test covering these cases (270 tests pass).

**Comment by robinebers** [comment]
That is an excellent addition. Thank you for both, letting us know in OpenUsage and submitting the PR here.

I hope this can be merged soon.

**Comment by robinebers** [comment]
@ryoppippi from my end too. this would be quite a nice addition.

**Comment by validatedev** [comment]
+++ looking forward to this being merged

**Comment by ryoppippi** [comment]
Closing as superseded by current main. The current Claude loader already has mtime-based filtering, usage-line marker scanning, and date-range filtering in apps/ccusage/src/adapter/claude/data-loader.ts, so this stale PR no longer needs to stay open.
